modules/core/include/opencv2/core/quaternion.hpp

```diff
@@ -0,0 +1,1949 @@
+
+    template <typename _Tp>
+    class Quat
+    {
+    using value_type = typename std::enable_if<
+        std::is_same<float, _Tp>::value||
+        std::is_same<double, _Tp>::value,
+        _Tp
+        >::type;
+    public:
+        Quat() = default;
+
+        /**
+         * @brief From Vec4d or Vec4f
+         */
+        explicit Quat(const cv::Vec<_Tp, 4> &coeff);
+
+        /**
+         * @brief from four numbers
+         */
+        Quat(_Tp w, _Tp x, _Tp y, _Tp z);
+
+        /**
+         * @brief from a angle, axis and qNorm with default 1, axis will be normalized in this constrctor. And
+         * it generates
+         * \f[q = [\cos\theta, u_x\sin\theta,u_y\sin\theta,  u_z\sin\theta].\f]
+         */
+        Quat(const _Tp angle, const cv::Vec<_Tp, 3> &axis);
+
+        /**
+         * @brief from a 3x3 rotate matrix
+         */
+        explicit Quat(const cv::Mat &R);
+
+        /**
+         * @brief from Rodrigues vector
+         */
+        explicit Quat(const Vec<_Tp, 3> &rvec);
+
+        /**
+         * @brief a way to get element
+         * @param index over a range [0, 3].
+         *
+         * Assume a quaternion q,
+         *
+         * q.at(0) is equicalent to q.w,
+         *
+         * q.at(1) is equicalent to q.x,
+         *
+         * q.at(2) is equicalent to q.y,
+         *
+         * q.at(3) is equicalent to q.z,
+         */
+        _Tp at(size_t index) const;
+
+        /**
+         * @brief return the conjugate of this quaternion
+         * \f[q.conjugate() = (w, -x, -y, -z)\f]
+         */
+        Quat<_Tp> conjugate() const;
+
+        /**
+         * @brief return the value of exponential value
+         * \f[\exp(q) = e^w (\cos||\boldsymbol{v}||+ \frac{v}{||\boldsymbol{v}||})\sin||\boldsymbol{v}||\f]
+         * @param q a quaternion
+         *
+         * For example:
+         * ```
+         * Quatd q{1,2,3,4};
+         * cout << exp(q) << endl;
+         * ```
+         */
+        template <typename T>
+        friend Quat<T> exp(const Quat<T> &q);
+
+        /**
+         * @brief return the value of exponential value
+         * \f[\exp(q) = e^w (\cos||\boldsymbol{v}||+ \frac{v}{||\boldsymbol{v}||}\sin||\boldsymbol{v}||)\f]
+         *
+         * For example
+         * ```
+         * Quatd q{1,2,3,4};
+         * cout << q.exp() << endl;
+         * ```
+         */
+        Quat<_Tp> exp() const;
+
+        /**
+         * @brief return the value of logarithm function
+         * \f[\ln(q) = \ln||q|| + \frac{\boldsymbol{v}}{||\boldsymbol{v}||}\arccos\frac{a}{||q||}\f]
+         * @param q a quaternion
+         * @param assumeUnit if true, q should be a unit quaternion and this function will save some computations.
+         *
+         * For example
+         * ```
+         * Quatd q1{1,2,3,4};
+         * cout << log(q1) << endl;
+         * ```
+         */
+        template <typename T>
+        friend Quat<T> log(const Quat<T> &q, bool assumeUnit);
```

Comment:
It can be implemented without friendship, because there is public method  `log`.